Lux/src/frgb.hpp

```diff
@@ -18,37 +18,47 @@
 
 typedef linalg::vec< float,3 > frgb;
 
-inline float r( const frgb &c );
-inline float g( const frgb &c );
-inline float b( const frgb &c );
+float r( const frgb &c );
+float g( const frgb &c );
+float b( const frgb &c );
 
 // returns single bytes per component - assumes [0.0, 1.0] range
 // clip or constrain out of range values before using
-inline unsigned char rc( const frgb &c );
-inline unsigned char gc( const frgb &c );
-inline unsigned char bc( const frgb &c );
+unsigned char rc( const frgb &c );
+unsigned char gc( const frgb &c );
+unsigned char bc( const frgb &c );
 
 // inline unsigned int ul( const frgb &c ) {} // bit shifty stuff
 
 // set component
-inline void setr( frgb &c, const float& r );
-inline void setg( frgb &c, const float& g );
-inline void setb( frgb &c, const float& b );
+void setr( frgb &c, const float& r );
+void setg( frgb &c, const float& g );
+void setb( frgb &c, const float& b );
 void set(  frgb &c, const float& r, const float& g, const float& b );
 frgb set(  const float& r, const float& g, const float& b );
 // TODO - set from bracketed list
 
-inline void setrc( frgb &c, const unsigned char& r );
-inline void setgc( frgb &c, const unsigned char& g );
-inline void setbc( frgb &c, const unsigned char& b );
+void setrc( frgb &c, const unsigned char& r );
+void setgc( frgb &c, const unsigned char& g );
+void setbc( frgb &c, const unsigned char& b );
 void setc(  frgb &c, const unsigned char& r, const unsigned char& g, const unsigned char& b );
 frgb setc(  const unsigned char& r, const unsigned char& g, const unsigned char& b );
 
 // I/O operators
 //std::ostream &operator << ( std::ostream &out, const frgb& f );
 void print_SRGB( const frgb &c );
 
-frgb& constrain( const frgb &c );   // Clip to range [ 0.0, 1.0 ] but keep colors in proportion
+// to clamp values use linalg::clamp()
+// component-wise clamp?
+void constrain( frgb &c );  // Clip to range [ 0.0, 1.0 ] but keep colors in proportion
+// frgb constrain( const frgb &c );   
 
+void apply_mask( const frgb &base, const frgb &mask, frgb &result );
```

Comment:
in C/C++ when only one parameter is mutating it is conventional to list it first not last as in strcpy

Reply:
this convention fits well with OOP too since the activating object this is passed as the first parameter in most OOP <-> free function conventions / compilers

Reply:
Fixed locally. Will update branch when I've worked through all the comments.